src/Control/Monad/Unify.hs

```diff
@@ -0,0 +1,244 @@
+  -- third argument with the outcome of that assertion; the computation inside
+  -- the callback argument will note the outcome of this assertion if either, or
+  -- both, arguments were fresh. If unification fails, the environment does not
+  -- change.
+  --
+  -- @since 1.0.0
+  unify ::
+    forall (a :: Type).
+    var ->
+    Either var def ->
+    (UnifyResult var def -> m a) ->
+    m a
+
+-- | @since 1.0.0
+instance (Eq def, Ord var, Monad m) => MonadUnify var def (UnifyT var def m) where
+  {-# INLINEABLE unify #-}
+  unify ::
+    forall (a :: Type).
+    var ->
+    Either var def ->
+    (UnifyResult var def -> UnifyT var def m a) ->
+    UnifyT var def m a
+  unify var1 y f = case y of
+    Left var2 -> do
+      defAbs1 <- UnifyT . asks $ definedAs var1
+      defAbs2 <- UnifyT . asks $ definedAs var2
+      case defAbs1 of
+        Nothing -> UnifyT $ case defAbs2 of
+          Nothing -> local (addNewEC var1 var2) (callback Equivalent)
+          Just (Left def2) -> local (define var1 def2) (callback . UnifiesTo $ def2)
+          Just (Right ec2) -> local (growEC var1 ec2) (callback Equivalent)
+        Just (Left def1) -> case defAbs2 of
+          Nothing -> UnifyT $ local (define var2 def1) (callback . UnifiesTo $ def1)
+          Just (Left def2) -> unifyConcrete def1 def2
+          Just (Right ec2) -> UnifyT $ local (defineEC ec2 def1) (callback . UnifiesTo $ def1)
+        Just (Right ec1) -> UnifyT $ case defAbs2 of
+          Nothing -> local (growEC var2 ec1) (callback Equivalent)
+          Just (Left def2) -> local (defineEC ec1 def2) (callback . UnifiesTo $ def2)
+          Just (Right ec2) -> local (unifyECs ec1 ec2) (callback Equivalent)
+    Right def -> handleMixed var1 def
+    where
+      handleMixed :: var -> def -> UnifyT var def m a
+      handleMixed var def =
+        (UnifyT . asks $ definedAs var) >>= \case
+          Nothing -> UnifyT $ local (define var def) (callback (UnifiesTo def))
+          Just (Left def') -> unifyConcrete def def'
+          Just (Right ec) -> UnifyT $ local (defineEC ec def) (callback (UnifiesTo def))
+      unifyConcrete :: def -> def -> UnifyT var def m a
+      unifyConcrete def1 def2 = f (if def1 == def2 then UnifiesTo def1 else DoesNotUnify var1 y)
+      callback :: UnifyResult var def -> ReaderT (UnifyEnv var def) m a
+      callback arg = let UnifyT comp = f arg in comp
+
+-- Helpers
+
+-- Internal unification state, which connects a variable to either a concrete
+-- definition, or to an equivalence class of other variables whose definition
+-- doesn't (yet) exist.
+newtype UnifyEnv (var :: Type) (def :: Type)
+  = UnifyEnv (Map var (Either def (NESet var)))
+  deriving (Eq) via (Map var (Either def (NESet var)))
+  deriving stock (Show)
+
+-- Helper for querying the state of a variable. It can either be fresh
+-- (`Nothing`), bound to a definition (`Just . Left`) or part of an equivalence
+-- class without a definition (`Just . Right`).
+definedAs ::
```

Comment:
Hmm, maybe we could use a custom sum type instead of  `Either` so it's more expressive? :thinking:

Reply:
Maybe even the `Maybe` could be replaced with something more specific.